File: todo-api-rs/src/application/use_cases/todo/find.rs
use crate::application::dtos::todo::find::{FindTodoError, FindTodoInput, FindTodoOutput};
use crate::application::repositories::todo::{FindError, TodoRepository};
use crate::domain::use_case::UseCase;

#[derive(Debug)]
pub struct FindTodoUseCase<T> {
    repository: T,
}

impl<T: TodoRepository> FindTodoUseCase<T> {
    pub fn new(repository: T) -> Self {
        Self { repository }
    }
}

impl<T: TodoRepository> UseCase<FindTodoInput, FindTodoOutput> for FindTodoUseCase<T> {
    async fn exec(self, todo_id: FindTodoInput) -> FindTodoOutput {
        self.repository
            .find(todo_id)
            .await
            .map_err(|err| match err {
                FindError::NotFound => FindTodoError::NotFound,
                FindError::Internal(err) => FindTodoError::Internal(err),
            })
    }
}

File: todo-api-js/src/application/use-cases/todo/find.ts
import { Err } from "@domain/utils/result";
import { InternalError } from "@domain/utils/exception";
import { type UseCase } from "@domain/use-case";
import { type TodoRepository } from "@application/repositories/todo";
import { type FindTodoInput, type FindTodoOutput } from "@application/dtos/todo/find";

export class FindTodoUseCase implements UseCase<FindTodoInput, FindTodoOutput> {
    constructor(private repository: TodoRepository) {}

    async exec(input: FindTodoInput): Promise<FindTodoOutput> {
        try {
            return this.repository.findById(input.id);
        } catch (e) {
            const error = new InternalError(e instanceof Error ? e : undefined);
            return new Err(error);
        }
    }
}
